### PR TITLE
VDI Example Terraform

### PR DESCRIPTION
**Pull request by jmarhee**
Provides an example for cloud users to:

- Deploy a virtual desktop through Terraform via module
- An example for consuming the Terraform module

Resolves https://github.com/coreweave/infra-pm/issues/84

### CODE REVIEW

**Comment by jmarhee** [comment]
This is re: https://github.com/coreweave/GTM-Requests/issues/82

**Comment by salanki** [comment]
Please [use a keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) in the description to link the PR to the [tracking issue](https://github.com/coreweave/infra-pm/issues/84). That way the issue gets auto closed and you can also see the PR when looking at the issue. I updated the description with this.

File: virtual-server/examples/terraform/vdi.tf [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  manifest = {
+    "apiVersion" = "virtualservers.coreweave.com/v1alpha1"
+    "kind"       = "VirtualServer"
+    "metadata" = {
+      "name"      = var.vdi_name
+      "namespace" = var.user_namespace
+    }
+    "spec" = {
+      "initializeRunning" = true
+      "network" = {
```
Comment (salanki) [inline virtual-server/examples/terraform/vdi.tf:13]: Does it make sense to make this section configurable by the user as variables?

File: virtual-server/examples/terraform/vdi.tf [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+          ]
+        }
+      }
+      "os" = {
+        "type" = var.vdi_os_type
+      }
+      "region" = var.vdi_region
+      "resources" = {
+        "cpu" = {
+          "count" = 3
```
Comment (salanki) [inline virtual-server/examples/terraform/vdi.tf:37]: This should be configurable just like GPU and RAM

File: virtual-server/examples/terraform/vdi.tf [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      "storage" = {
+        "root" = {
+          "size" = var.vdi_root_storage
+          "source" = {
+            "pvc" = {
+              "name"      = var.vdi_image
+              "namespace" = "vd-images"
+            }
+          }
+          "storageClassName" = "block-nvme-ord1"
```
Comment (salanki) [inline virtual-server/examples/terraform/vdi.tf:54]: The storage class name should be derived from the region. Ie, pseudocode: `block-nvme-${var.vdi_region}`.

**Review by salanki — CHANGES_REQUESTED** [review]

**Comment by salanki** [comment]
This is neat. I can see how this would evolve to Terraform modules provided by us.

File: virtual-server/examples/terraform/README.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# Deploying VirtualServers to Kubernetes with Terraform
+
+This [terraform](terraform.io) module uses the [kubernetes provider](https://registry.terraform.io/providers/hashicorp/kubernetes/latest/docs) to deploy `VirtualServers` (Virtual Desktops) to [CoreWeave Cloud](coreweave.io).
```
[the remark this answers is not in the document]
Comment (brianventuro) [reply virtual-server/examples/terraform/README.md:3]: coreweave.com

File: virtual-server/examples/terraform/variables.tf [hunk 5]
```diff
@@ -0,0 +1,89 @@
+variable "kubeconfig_path" {
+  description = "Path to kubeconfig"
+  default     = "~/.kube/config"
+}
+
+variable "user_namespace" {
+  description = "Namespace VDI will be installed to"
```
Comment (salanki) [inline virtual-server/examples/terraform/variables.tf:7]: Change to Virtual Server

File: virtual-server/examples/terraform/variables.tf [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+variable "vs_name" {
+  description = "VDI hostname"
+  default     = "MY-VDI"
```
Comment (salanki) [inline virtual-server/examples/terraform/variables.tf:12]: VS

File: virtual-server/examples/terraform/variables.tf [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+variable "vs_username" {
+  description = "VDI username"
```
Comment (salanki) [inline virtual-server/examples/terraform/variables.tf:16]: VS

File: virtual-server/examples/terraform/variables.tf [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+variable "vs_generate_password" {
+  type        = bool
+  default     = true
+  description = "Generate password"
+}
+
+variable "vs_memory" {
+  description = "VDI RAM"
```
Comment (salanki) [inline virtual-server/examples/terraform/variables.tf:32]: VS

File: virtual-server/examples/terraform/variables.tf [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  default     = "16Gi"
+}
+
+variable "vs_root_storage" {
+  description = "VDI root device storage (i.e 80Gi)"
```
Comment (salanki) [inline virtual-server/examples/terraform/variables.tf:37]: VS

File: virtual-server/examples/terraform/vs.tf [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  manifest = {
+    "apiVersion" = "virtualservers.coreweave.com/v1alpha1"
+    "kind"       = "VirtualServer"
+    "metadata" = {
+      "name"      = var.vs_name
+      "namespace" = var.user_namespace
+    }
+    "spec" = {
+      "initializeRunning" = var.vs_running
+      "network" = {
```
Comment (salanki) [inline virtual-server/examples/terraform/vs.tf:13]: Can we add in attachLoadBalancerIP here as well. It gets the Service IP directly attached to the VM (it becomes the Pod IP). It's a pretty common use case so good for completeness.

**Review by salanki — CHANGES_REQUESTED** [review]
Some minor tweaks left for consistency on Virtual Server / VS. Once this is merged, please add a description and link this repo as a new entry under "Deployment Methods" in the documentation, with some CLI examples how to run it. You can access gitbook where we edit our docs via login.CoreWeave.com

**Review by salanki — APPROVED** [review]
@jmarhee: Can you give me some example output how this looks on the CLI?

**Comment by jmarhee** [comment]
> @jmarhee: Can you give me some example output how this looks on the CLI?

The Kubernetes provider requires the attributes this module outputs to be marked sensitive for CRDs via this resource type, so I can't print them out to STDOUT (but are exported so another module can consume them) but what I can do is update the output to add a output message/command for the client to get their password from the state (and the IP from `kubectl`) -- would that be sufficient here (and update the documentation to reflect that as well)?

**Comment by salanki** [comment]
So there is something in the Kubernetes provider limiting you to fetch the IP address from the VS status? Can you explain that more, seems unreasonable, no way to disable? If we need to instruct users, they should be doing `kubectl get virtualserver` not go after the service directly.

**Comment by jmarhee** [comment]
Yes, this is being enforced by at the provider level, [as I understood it](https://www.terraform.io/docs/language/values/outputs.html#sensitive-suppressing-values-in-cli-output), for this resource type's exports, but can confirm. 

I can access it via the output CLI command since that reads from the state directly where it is in clear text; I can definitely look into un-enforcing this, because that would be much more useful for a case like this where we want users to have the information they need without going back into kubectl to get it (or I can just direct users back to the virtualserver resource in kubectl to get, both, the IP and credentials, etc. etc.).

**Comment by jmarhee** [comment]
Actually, I have another method I can try to get the information through Terraform (from the provider, rather than try to export something from the resource), I'll try that first, and then see about unenforcing the sensitive value rules.

**Comment by jmarhee** [comment]
The last push includes the above-mentioned workaround for retrieving the VS password, and extracting the Service IP from the Kubernetes provider, sidestepping the provider-enforced resource attributes. 

```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

vs_network = "[address]"
vs_password = "[password]"
```

**Comment by salanki** [comment]
Awesome. I'm gonna merge it, please write the docs in gitbook and close the tracking ticket once that is done!

**Pull request merged**